iwyu.cc

```diff
@@ -2334,6 +2334,149 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
+          for (const TypeSourceInfo* arg : drop_begin(expr->getArgs())) {
+            const QualType arg_type = arg->getType();
+            // Conversion functions of class/struct/union types are not trivial.
+            if (RemoveReference(arg_type)->isRecordType())
+              return true;
+            if (!IsConvertible(arg_type, elem_type, CurrentLoc(),
+                               compiler()->getSema())) {
+              return true;
+            }
+          }
+          // Report types.
+          for (const TypeSourceInfo* arg : drop_begin(expr->getArgs())) {
+            const QualType arg_type = arg->getType();
+            const QualType arg_deref_type = RemoveReference(arg_type);
+            if (elem_type->isPointerType()) {
+              // No full type needed if the pointed-to types are the same.
+              const Type* arg_pointee_type = GetCanonicalType(
+                  arg_deref_type->getPointeeOrArrayElementType());
+              const Type* elem_pointee_type =
+                  GetCanonicalType(elem_type->getPointeeType().getTypePtr());
+              if (arg_pointee_type == elem_pointee_type)
+                continue;
+              ReportTypeUse(CurrentLoc(), arg_type.getTypePtr(),
+                            DerefKind::RemoveRefsAndPtr);
+            } else if (const auto* elem_mem_ptr_type =
+                           elem_type->getAs<MemberPointerType>()) {
+              const Type* elem_qualifier =
+                  elem_mem_ptr_type->getQualifier()->getAsType();
+              if (const auto* arg_mem_ptr_type =
+                      arg_deref_type->getAs<MemberPointerType>()) {
+                const Type* arg_class = GetCanonicalType(
+                    arg_mem_ptr_type->getQualifier()->getAsType());
+                if (arg_class != GetCanonicalType(elem_qualifier)) {
+                  ReportTypeUse(CurrentLoc(), elem_qualifier, DerefKind::None);
+                  return true;
+                }
+              }
+            }
+          }
+          return true;
+        }
+
+        // The remaining cases involve exactly 2 arguments.
+        if (expr->getNumArgs() != 2)
+          return true;
+        const QualType lhs_deref_type = RemoveReference(lhs_qual_type);
+        const QualType rhs_deref_type = RemoveReference(rhs_qual_type);
+        if (lhs_deref_type->isStructureOrClassType() &&
+            rhs_deref_type->isStructureOrClassType()) {
+          // Direct reference binding cases. See [dcl.init.ref].
+          // Only (maybe, potential) derived-to-base binding cases should be
+          // taken into account.
+          if (GetCanonicalType(lhs_deref_type.getTypePtr()) ==
+              GetCanonicalType(rhs_deref_type.getTypePtr())) {
+            return true;
+          }
+          const bool lvalue_bindable = lhs_type->isLValueReferenceType() &&
+                                       rhs_type->isLValueReferenceType();
+          const bool rvalue_bindable =
```

Comment:
I think you can drop these `const` for locals.

Reply:
Actually, maybe I'm internally inconsistent :-) -- I think `const T*` looks perfectly normal but `const T` for a value type looks excessive. So to follow that logic, maybe it makes sense to drop `const` for the `QualType` locals above too.

Reply:
Yeah, I don't have strong feelings either way, it just looks a little noisy to me.